File: distribution/npm/src/core/diff.ts
/**
 * Incremental scene diffing with stable element IDs
 * 
 * Uses content-addressed hashing + ID-based reconciliation for O(n) diffing
 * with minimal SVG regeneration. Mirrors Rust implementation for cross-platform parity.
 */

import type { Shape, Style, Transform } from './types';

// ─────────────────────────────────────────────────────────────────────────────
// Hashing
// ─────────────────────────────────────────────────────────────────────────────

const FNV_OFFSET = 0xcbf29ce484222325n;
const FNV_PRIME = 0x100000001b3n;

/** FNV-1a hash function for stable identity */
export function fnv1a(data: string): bigint {
  let hash = FNV_OFFSET;
  for (let i = 0; i < data.length; i++) {
    hash ^= BigInt(data.charCodeAt(i));
    hash = BigInt.asUintN(64, hash * FNV_PRIME);
  }
  return hash;
}

/** Content hash for element equality */
export type ContentHash = bigint;

/** Stable element identity */
export type ElementId = bigint;

/** Compute stable ID from creation order and kind */
export function computeId(order: number, kind: string, keyProps?: Record<string, unknown>): ElementId {
  let key = `${order}:${kind}`;
  if (keyProps) {
    const sorted = Object.keys(keyProps).sort();
    for (const k of sorted) key += `:${k}=${JSON.stringify(keyProps[k])}`;
  }
  return fnv1a(key);
}

// ─────────────────────────────────────────────────────────────────────────────
// Indexed Elements
// ─────────────────────────────────────────────────────────────────────────────

export interface IndexedElement {
  id: ElementId;
  hash: ContentHash;
  kind: string;
  index: number;
  svg: string;
}

/** Extract key properties for identity computation */
function keyProps(shape: Shape): Record<string, unknown> {
  const { kind, props } = shape;
  switch (kind) {
    case 'rect': return { at: props.at, size: props.size };
    case 'circle': return { at: props.at, radius: props.radius };
    case 'ellipse': return { at: props.at, size: props.size };
    case 'line': return { from: props.from, to: props.to };
    case 'path': return { d: props.d ?? props.content };
    case 'polygon': return { points: props.points };
    case 'text': return { at: props.at, content: props.content };
    case 'image': return { href: props.href };
    default: return {};
  }
}

// ─────────────────────────────────────────────────────────────────────────────
// Diff Operations
// ─────────────────────────────────────────────────────────────────────────────

export type DiffOp =
  | { type: 'none' }
  | { type: 'full_redraw' }
  | { type: 'add'; id: string; idx: number; svg: string }
  | { type: 'remove'; id: string; idx: number }
  | { type: 'update'; id: string; idx: number; attrs: [string, string][]; svg?: string }
  | { type: 'move'; id: string; from: number; to: number }
  | { type: 'update_defs'; svg: string };

// ─────────────────────────────────────────────────────────────────────────────
// Scene Indexing
// ─────────────────────────────────────────────────────────────────────────────

export interface SceneIndex {
  elements: IndexedElement[];
  idMap: Map<string, number>;
}

export interface IndexableScene {
  canvas: { width: number; height: number; fill: string };
  shapes: ShapeDict[];
  defs: string[];
}

interface ShapeDict {
  kind: string;
  props: Record<string, unknown>;
  style: Record<string, unknown>;
  transform: Record<string, unknown>;
  children: ShapeDict[];
}

/** Index a scene for O(1) lookups */
export function indexScene(scene: IndexableScene, renderShape: (s: ShapeDict) => string): SceneIndex {
  const elements: IndexedElement[] = [];
  const idMap = new Map<string, number>();

  for (let i = 0; i < scene.shapes.length; i++) {
    const shape = scene.shapes[i];
    const svg = renderShape(shape);
    const id = computeId(i, shape.kind, shape.props as Record<string, unknown>);
    const hash = fnv1a(svg);
    const idStr = id.toString(16);

    elements.push({ id, hash, kind: shape.kind, index: i, svg });
    idMap.set(idStr, i);
  }

  return { elements, idMap };
}

// ─────────────────────────────────────────────────────────────────────────────
// Diffing
// ─────────────────────────────────────────────────────────────────────────────

export interface DiffResult {
  ops: DiffOp[];
  canvasChanged: boolean;
}

/** Diff two indexed scenes */
export function diff(old: SceneIndex, oldScene: IndexableScene, newScene: IndexableScene, renderShape: (s: ShapeDict) => string): DiffResult {
  // Canvas change = full redraw
  if (
    oldScene.canvas.width !== newScene.canvas.width ||
    oldScene.canvas.height !== newScene.canvas.height ||
    oldScene.canvas.fill !== newScene.canvas.fill
  ) {
    return { ops: [{ type: 'full_redraw' }], canvasChanged: true };
  }

  // Fast path: empty scenes
  if (old.elements.length === 0 && newScene.shapes.length === 0) {
    return { ops: [], canvasChanged: false };
  }

  const ops: DiffOp[] = [];
  const matched = new Array(old.elements.length).fill(false);

  // Pass 1: Match new elements to old
  for (let newIdx = 0; newIdx < newScene.shapes.length; newIdx++) {
    const shape = newScene.shapes[newIdx];
    const newId = computeId(newIdx, shape.kind, shape.props as Record<string, unknown>);
    const newIdStr = newId.toString(16);
    const svg = renderShape(shape);
    const newHash = fnv1a(svg);

    const oldIdx = old.idMap.get(newIdStr);

    if (oldIdx !== undefined) {
      matched[oldIdx] = true;
      const oldEl = old.elements[oldIdx];

      // Content changed
      if (oldEl.hash !== newHash) {
        const attrs = diffAttrs(oldScene.shapes[oldIdx], shape);
        ops.push({
          type: 'update',
          id: newIdStr,
          idx: newIdx,
          attrs,
          svg: attrs.length > 3 ? svg : undefined,
        });
      }

      // Position changed
      if (oldIdx !== newIdx) {
        ops.push({ type: 'move', id: newIdStr, from: oldIdx, to: newIdx });
      }
    } else {
      // New element
      ops.push({ type: 'add', id: newIdStr, idx: newIdx, svg });
    }
  }

  // Pass 2: Remove unmatched old elements (reverse for stable indices)
  for (let i = matched.length - 1; i >= 0; i--) {
    if (!matched[i]) {
      const oldEl = old.elements[i];
      ops.push({ type: 'remove', id: oldEl.id.toString(16), idx: i });
    }
  }

  // Check defs changes
  const oldDefs = oldScene.defs.join('');
  const newDefs = newScene.defs.join('');
  if (oldDefs !== newDefs) {
    ops.push({ type: 'update_defs', svg: newDefs });
  }

  return { ops, canvasChanged: false };
}

// ─────────────────────────────────────────────────────────────────────────────
// Attribute Diffing
// ─────────────────────────────────────────────────────────────────────────────

function diffAttrs(old: ShapeDict, newShape: ShapeDict): [string, string][] {
  const changes: [string, string][] = [];

  // Props diff
  for (const [k, v] of Object.entries(newShape.props)) {
    if (JSON.stringify(old.props[k]) !== JSON.stringify(v)) {
      changes.push([propToAttr(k), formatValue(v)]);
    }
  }

  // Style diff
  for (const [k, v] of Object.entries(newShape.style)) {
    if (JSON.stringify(old.style[k]) !== JSON.stringify(v)) {
      changes.push([styleToAttr(k), formatValue(v)]);
    }
  }

  // Transform diff
  if (JSON.stringify(old.transform) !== JSON.stringify(newShape.transform)) {
    changes.push(['transform', buildTransform(newShape.transform)]);
  }

  return changes;
}

function propToAttr(prop: string): string {
  switch (prop) {
    case 'at': return 'x';
    case 'size': return 'width';
    case 'radius': return 'r';
    case 'content': return 'textContent';
    default: return prop;
  }
}

function styleToAttr(style: string): string {
  switch (style) {
    case 'strokeWidth': return 'stroke-width';
    case 'fontSize': return 'font-size';
    case 'fontWeight': return 'font-weight';
    case 'textAnchor': return 'text-anchor';
    default: return style;
  }
}

function formatValue(v: unknown): string {
  if (Array.isArray(v)) return v.join(',');
  return String(v ?? '');
}

function buildTransform(tf: Record<string, unknown>): string {
  const parts: string[] = [];
  if (tf.translate) {
    const [tx, ty] = tf.translate as [number, number];
    parts.push(`translate(${tx},${ty})`);
  }
  if (tf.rotate) {
    parts.push(`rotate(${tf.rotate})`);
  }
  if (tf.scale) {
    const [sx, sy] = tf.scale as [number, number];
    parts.push(`scale(${sx},${sy})`);
  }
  return parts.join(' ');
}

// ─────────────────────────────────────────────────────────────────────────────
// Render Cache
// ─────────────────────────────────────────────────────────────────────────────

interface CacheEntry {
  svg: string;
  hits: number;
}

/** SVG fragment memoization cache */
export class RenderCache {
  private entries = new Map<string, CacheEntry>();
  private maxSize: number;

  constructor(maxSize = 1024) {
    this.maxSize = maxSize;
  }

  get(hash: ContentHash): string | undefined {
    const key = hash.toString(16);
    const entry = this.entries.get(key);
    if (entry) {
      entry.hits++;
      return entry.svg;
    }
    return undefined;
  }

  set(hash: ContentHash, svg: string): void {
    if (this.entries.size >= this.maxSize) this.evictLru();
    this.entries.set(hash.toString(16), { svg, hits: 1 });
  }

  getOrCompute(hash: ContentHash, compute: () => string): string {
    const cached = this.get(hash);
    if (cached !== undefined) return cached;
    const svg = compute();
    this.set(hash, svg);
    return svg;
  }

  private evictLru(): void {
    let minKey: string | undefined;
    let minHits = Infinity;
    for (const [key, entry] of this.entries) {
      if (entry.hits < minHits) {
        minHits = entry.hits;
        minKey = key;
      }
    }
    if (minKey) this.entries.delete(minKey);
  }

  clear(): void {
    this.entries.clear();
  }

  get size(): number {
    return this.entries.size;
  }
}

File: source/core/src/render/diff.rs
//! Incremental scene diffing with stable element IDs
//!
//! Uses content-addressed hashing + ID-based reconciliation for O(n) diffing
//! with minimal SVG regeneration. Inspired by VDOM reconciliation algorithms.

use std::collections::HashMap;
use crate::hash::{ContentHash, ElementId, ElementKind, Fnv1a, IdGen};
use crate::scene::{Element, Scene, Style};

/// Indexed element with stable identity and content hash
#[derive(Debug, Clone)]
pub struct IndexedElement {
    pub id: ElementId,
    pub hash: ContentHash,
    pub kind: ElementKind,
    pub index: usize,
}

impl IndexedElement {
    pub fn new(el: &Element, order: u64, index: usize) -> Self {
        let kind = element_kind(el);
        let id = compute_id(el, order, kind);
        let hash = ContentHash::from_svg(&el.to_svg());
        Self { id, hash, kind, index }
    }
}

/// Compute stable ID from element's key properties
fn compute_id(el: &Element, order: u64, kind: ElementKind) -> ElementId {
    let mut h = Fnv1a::default();
    
    match el {
        Element::Rect(r) => { h.write_f32(r.x); h.write_f32(r.y); }
        Element::Circle(c) => { h.write_f32(c.cx); h.write_f32(c.cy); }
        Element::Ellipse(e) => { h.write_f32(e.cx); h.write_f32(e.cy); }
        Element::Line(l) => { h.write_f32(l.x1); h.write_f32(l.y1); h.write_f32(l.x2); h.write_f32(l.y2); }
        Element::Path(p) => h.write_str(&p.d),
        Element::Polygon(p) => for (x, y) in &p.points { h.write_f32(*x); h.write_f32(*y); },
        Element::Text(t) => { h.write_f32(t.x); h.write_f32(t.y); h.write_str(&t.content); }
        Element::Image(i) => { h.write_str(&i.href); }
        Element::Group(_, tf) => if let Some(t) = tf { h.write_str(t); },
    }
    
    ElementId::with_key(order, kind.as_u8(), &h.finish().to_le_bytes())
}

/// Get element kind discriminant
#[inline]
pub fn element_kind(el: &Element) -> ElementKind {
    match el {
        Element::Rect(_) => ElementKind::Rect,
        Element::Circle(_) => ElementKind::Circle,
        Element::Ellipse(_) => ElementKind::Ellipse,
        Element::Line(_) => ElementKind::Line,
        Element::Path(_) => ElementKind::Path,
        Element::Polygon(_) => ElementKind::Polygon,
        Element::Text(_) => ElementKind::Text,
        Element::Image(_) => ElementKind::Image,
        Element::Group(_, _) => ElementKind::Group,
    }
}

/// Targeted diff operation for incremental updates
#[derive(Debug, Clone, PartialEq)]
pub enum DiffOp {
    None,
    FullRedraw,
    Add { id: u64, idx: usize, svg: String },
    Remove { id: u64, idx: usize },
    Update { id: u64, idx: usize, attrs: Vec<(String, String)>, svg: Option<String> },
    Move { id: u64, from: usize, to: usize },
    UpdateDefs { svg: String },
}

/// Indexed scene for O(1) element lookup
#[derive(Debug, Default)]
pub struct IndexedScene {
    pub elements: Vec<IndexedElement>,
    id_map: HashMap<ElementId, usize>,
}

impl IndexedScene {
    pub fn from_scene(scene: &Scene) -> Self {
        let gen = IdGen::default();
        let elements: Vec<_> = scene.elements()
            .iter()
            .enumerate()
            .map(|(idx, el)| IndexedElement::new(el, gen.next(), idx))
            .collect();
        
        let id_map = elements.iter().map(|e| (e.id, e.index)).collect();
        Self { elements, id_map }
    }

    #[inline]
    pub fn get(&self, id: &ElementId) -> Option<&IndexedElement> {
        self.id_map.get(id).map(|&idx| &self.elements[idx])
    }

    #[inline]
    pub fn len(&self) -> usize { self.elements.len() }

    #[inline]
    pub fn is_empty(&self) -> bool { self.elements.is_empty() }
}

/// Diff result with operations
#[derive(Debug, Default)]
pub struct DiffResult {
    pub ops: Vec<DiffOp>,
    pub canvas_changed: bool,
}

impl DiffResult {
    pub fn full_redraw() -> Self {
        Self { ops: vec![DiffOp::FullRedraw], canvas_changed: true }
    }

    pub fn empty() -> Self { Self::default() }

    #[inline]
    pub fn is_empty(&self) -> bool { self.ops.is_empty() }

    #[inline]
    pub fn needs_full_redraw(&self) -> bool {
        self.canvas_changed || self.ops.iter().any(|o| matches!(o, DiffOp::FullRedraw))
    }
}

/// Diff two scenes using indexed reconciliation
pub fn diff(old: &Scene, new: &Scene) -> DiffResult {
    if old.width != new.width || old.height != new.height || old.background != new.background {
        return DiffResult::full_redraw();
    }

    let old_els = old.elements();
    let new_els = new.elements();

    if old_els.is_empty() && new_els.is_empty() {
        return DiffResult::empty();
    }

    let old_indexed = IndexedScene::from_scene(old);
    let gen = IdGen::default();
    
    let mut ops = Vec::new();
    let mut matched: Vec<bool> = vec![false; old_els.len()];

    for (new_idx, new_el) in new_els.iter().enumerate() {
        let new_kind = element_kind(new_el);
        let new_id = compute_id(new_el, gen.next(), new_kind);
        let new_hash = ContentHash::from_svg(&new_el.to_svg());

        if let Some(old_ie) = old_indexed.get(&new_id) {
            matched[old_ie.index] = true;
            
            if old_ie.hash != new_hash {
                let attrs = diff_attrs(&old_els[old_ie.index], new_el);
                let svg = if attrs.len() > 3 { Some(new_el.to_svg()) } else { None };
                ops.push(DiffOp::Update { id: new_id.0, idx: new_idx, attrs, svg });
            }
            
            if old_ie.index != new_idx {
                ops.push(DiffOp::Move { id: new_id.0, from: old_ie.index, to: new_idx });
            }
        } else {
            ops.push(DiffOp::Add { id: new_id.0, idx: new_idx, svg: new_el.to_svg() });
        }
    }

    for (old_idx, &was_matched) in matched.iter().enumerate().rev() {
        if !was_matched {
            let old_el = &old_els[old_idx];
            let old_kind = element_kind(old_el);
            let old_id = compute_id(old_el, old_idx as u64, old_kind);
            ops.push(DiffOp::Remove { id: old_id.0, idx: old_idx });
        }
    }

    let old_defs = build_defs_svg(old);
    let new_defs = build_defs_svg(new);
    if old_defs != new_defs {
        ops.push(DiffOp::UpdateDefs { svg: new_defs });
    }

    DiffResult { ops, canvas_changed: false }
}

fn build_defs_svg(scene: &Scene) -> String {
    let mut svg = String::new();
    for g in scene.gradients() { svg.push_str(&g.to_svg()); }
    for f in scene.filters() { svg.push_str(&f.to_svg()); }
    svg
}

fn diff_attrs(old: &Element, new: &Element) -> Vec<(String, String)> {
    let mut changes = Vec::new();

    match (old, new) {
        (Element::Rect(o), Element::Rect(n)) => {
            if o.x != n.x { changes.push(("x".into(), n.x.to_string())); }
            if o.y != n.y { changes.push(("y".into(), n.y.to_string())); }
            if o.w != n.w { changes.push(("width".into(), n.w.to_string())); }
            if o.h != n.h { changes.push(("height".into(), n.h.to_string())); }
            if o.rx != n.rx { changes.push(("rx".into(), n.rx.to_string())); }
            diff_style(&o.style, &n.style, &mut changes);
            diff_transform(&o.transform, &n.transform, &mut changes);
        }
        (Element::Circle(o), Element::Circle(n)) => {
            if o.cx != n.cx { changes.push(("cx".into(), n.cx.to_string())); }
            if o.cy != n.cy { changes.push(("cy".into(), n.cy.to_string())); }
            if o.r != n.r { changes.push(("r".into(), n.r.to_string())); }
            diff_style(&o.style, &n.style, &mut changes);
            diff_transform(&o.transform, &n.transform, &mut changes);
        }
        (Element::Ellipse(o), Element::Ellipse(n)) => {
            if o.cx != n.cx { changes.push(("cx".into(), n.cx.to_string())); }
            if o.cy != n.cy { changes.push(("cy".into(), n.cy.to_string())); }
            if o.rx != n.rx { changes.push(("rx".into(), n.rx.to_string())); }
            if o.ry != n.ry { changes.push(("ry".into(), n.ry.to_string())); }
            diff_style(&o.style, &n.style, &mut changes);
            diff_transform(&o.transform, &n.transform, &mut changes);
        }
        (Element::Line(o), Element::Line(n)) => {
            if o.x1 != n.x1 { changes.push(("x1".into(), n.x1.to_string())); }
            if o.y1 != n.y1 { changes.push(("y1".into(), n.y1.to_string())); }
            if o.x2 != n.x2 { changes.push(("x2".into(), n.x2.to_string())); }
            if o.y2 != n.y2 { changes.push(("y2".into(), n.y2.to_string())); }
            diff_style(&o.style, &n.style, &mut changes);
            diff_transform(&o.transform, &n.transform, &mut changes);
        }
        (Element::Path(o), Element::Path(n)) => {
            if o.d != n.d { changes.push(("d".into(), n.d.clone())); }
            diff_style(&o.style, &n.style, &mut changes);
            diff_transform(&o.transform, &n.transform, &mut changes);
        }
        (Element::Text(o), Element::Text(n)) => {
            if o.x != n.x { changes.push(("x".into(), n.x.to_string())); }
            if o.y != n.y { changes.push(("y".into(), n.y.to_string())); }
            if o.content != n.content { changes.push(("textContent".into(), n.content.clone())); }
            if o.font != n.font { changes.push(("font-family".into(), n.font.clone())); }
            if o.size != n.size { changes.push(("font-size".into(), n.size.to_string())); }
            if o.weight != n.weight { changes.push(("font-weight".into(), n.weight.clone())); }
            if o.anchor != n.anchor { changes.push(("text-anchor".into(), n.anchor.clone())); }
            diff_style(&o.style, &n.style, &mut changes);
            diff_transform(&o.transform, &n.transform, &mut changes);
        }
        (Element::Image(o), Element::Image(n)) => {
            if o.x != n.x { changes.push(("x".into(), n.x.to_string())); }
            if o.y != n.y { changes.push(("y".into(), n.y.to_string())); }
            if o.w != n.w { changes.push(("width".into(), n.w.to_string())); }
            if o.h != n.h { changes.push(("height".into(), n.h.to_string())); }
            if o.href != n.href { changes.push(("href".into(), n.href.clone())); }
            diff_transform(&o.transform, &n.transform, &mut changes);
        }
        (Element::Polygon(o), Element::Polygon(n)) => {
            if o.points != n.points {
                let pts: String = n.points.iter()
                    .map(|(x, y)| format!("{},{}", x, y))
                    .collect::<Vec<_>>()
                    .join(" ");
                changes.push(("points".into(), pts));
            }
            diff_style(&o.style, &n.style, &mut changes);
            diff_transform(&o.transform, &n.transform, &mut changes);
        }
        _ => {}
    }

    changes
}

fn diff_style(old: &Style, new: &Style, out: &mut Vec<(String, String)>) {
    if old.fill != new.fill {
        out.push(("fill".into(), new.fill.clone().unwrap_or_default()));
    }
    if old.stroke != new.stroke {
        out.push(("stroke".into(), new.stroke.clone().unwrap_or_default()));
    }
    if old.stroke_width != new.stroke_width {
        out.push(("stroke-width".into(), new.stroke_width.to_string()));
    }
    if old.opacity != new.opacity {
        out.push(("opacity".into(), new.opacity.to_string()));
    }
    if old.filter != new.filter {
        let val = new.filter.as_ref().map(|f| format!("url(#{})", f)).unwrap_or_default();
        out.push(("filter".into(), val));
    }
}

#[inline]
fn diff_transform(old: &Option<String>, new: &Option<String>, out: &mut Vec<(String, String)>) {
    if old != new {
        out.push(("transform".into(), new.clone().unwrap_or_default()));
    }
}

pub type Patch = DiffOp;

#[cfg(test)]
mod tests {
    use super::*;
    use crate::scene::{Circle, Rect};

    fn make_scene(w: u32, h: u32, bg: &str) -> Scene {
        Scene::new_internal(w, h, bg.to_string())
    }

    #[test]
    fn test_diff_result_empty() {
        let r = DiffResult::empty();
        assert!(r.is_empty());
        assert!(!r.needs_full_redraw());
    }

    #[test]
    fn test_diff_result_full_redraw() {
        let r = DiffResult::full_redraw();
        assert!(r.needs_full_redraw());
        assert!(r.canvas_changed);
    }

    #[test]
    fn test_identical_scenes() {
        let s1 = make_scene(800, 600, "#fff");
        let s2 = make_scene(800, 600, "#fff");
        assert!(diff(&s1, &s2).is_empty());
    }

    #[test]
    fn test_canvas_change_triggers_redraw() {
        let s1 = make_scene(800, 600, "#fff");
        let s2 = make_scene(1024, 600, "#fff");
        assert!(diff(&s1, &s2).needs_full_redraw());
    }

    #[test]
    fn test_indexed_scene_empty() {
        let scene = make_scene(800, 600, "#fff");
        let indexed = IndexedScene::from_scene(&scene);
        assert!(indexed.is_empty());
    }

    #[test]
    fn test_element_kind_rect() {
        let el = Element::Rect(Rect { x: 0.0, y: 0.0, w: 100.0, h: 50.0, rx: 0.0, style: Style::default(), transform: None });
        assert_eq!(element_kind(&el), ElementKind::Rect);
    }

    #[test]
    fn test_element_kind_circle() {
        let el = Element::Circle(Circle { cx: 50.0, cy: 50.0, r: 25.0, style: Style::default(), transform: None });
        assert_eq!(element_kind(&el), ElementKind::Circle);
    }
}
